galaxyproject/galaxy
# Add `email` notifications channel

Closes #17882

This PR adds a new `email` channel to the notifications system.

## Backend changes
- Includes 2 database migrations to add 2 new columns to the `notification` table.
  - `dispatched` column: will determine if this notification has been "dispatched" through the additional communication channels. Any existing notification will set this value to "True" so users won't get emails retroactively.
  - `galaxy_url` column: allows to build URLs from the notification channel plugins that don't have access to the web request context.
- Move the creation of notifications (database objects) to a celery task if enabled. This can offload resolving the recipient users when we include complex combinations of groups and/or roles as recipients. We can also control if we want to explicitly run this synchronously when we don't use groups or roles.
- Adds a new periodic celery task (every 10 minutes by default) that will process non `dispatched` notifications and send them via all available channels. In-app notifications don't depend on this task so they will pop up in Galaxy at their due time. Other external channels, like email, will probably carry some delay depending on the `dispatch_notifications_interval` configuration option.
- Adds personalizable email templates for the currently existing types of notifications.

## Frontend changes
- The notification preferences will display a new `Email` channel if the server "supports" it. The channels supported by the server are passed to the client via a new header named `supported-channels` when requesting `api/notifications/preferences`. 
- Navigating to `/user/notifications?preferences=true` will display the preferences panel open. This makes it easier to link to the preferences from the email templates.

## Sample Emails

### Shared Item
![Screenshot from 2024-04-11 13-46-29](https://github.com/galaxyproject/galaxy/assets/46503462/b907b1c2-99b1-4b17-afb2-4e6c49451e7d)


### Direct Message
![Screenshot from 2024-04-11 13-49-10](https://github.com/galaxyproject/galaxy/assets/46503462/4b36b2dd-6cad-4668-b680-4c1ec2e32828)

## See it in action
![NotificationEmails](https://github.com/galaxyproject/galaxy/assets/46503462/f982f132-5ce0-43b7-bc67-89a91c7bb7d1)


## How to test the changes?
  - Make sure `enable_celery_tasks` is set to true in your Galaxy config.
  - Set up the `smtp_server` in your config. I used https://github.com/rnwood/smtp4dev/
  - Play around with the email templates from `lib/galaxy/config/templates/mail/notifications/` by copying them and setting the `templates_dir` accordingly.
  - Play around disabling the email channel for certain notifications in your preferences.
  - Play around scheduling notifications as an admin.

## License
